// abbrev/show.js
import React, {Component} from 'react';
import Layout from '../../componenets/Layout';
import Contract from '../../ethereum/contract';
import { Card, Grid, Button } from 'semantic-ui-react';
import web3 from '../../ethereum/web3';
import {Link} from '../../routes';

class CampaignShow extends Component {

  static async getInitialProps (props){
    const contract = Contract(props.query.address);

    const summary = await contract.methods.getContractSummary().call();
    console.log(props.query.address);

    return{
      address: props.query.address,
      title: summary[6],
      };
    }


renderCards(){
  const {
  title
  } = this.props;

  const items =[
  {
    header: title,
    meta: 'Minium Contribution (wei)',
    description: 'You must contribute at least this much wei',

  },

];
    return <Card.Group items ={items} />;
  }



  render (){
    return(
      <Layout>
      <h3> Contract  </h3>
        <Grid>
          <Grid.Row>
            <Grid.Column width ={10}>
              {this.renderCards()}
            </Grid.Column>

            <Grid.Column width = {6}>

            </Grid.Column>
         </Grid.Row>

     </Grid>
    </Layout>
   );

  }
}

export default CampaignShow;
